### PR TITLE
Make ResourcesPlugin more dynamic and better handling early start-up

### PR DESCRIPTION
**Pull request by laeubi**
Fix https://github.com/eclipse-platform/eclipse.platform.resources/issues/52

This finally delays workspace creation until the INSTANCE location is properly set (aka Workspace Selection Dialog) and even if the ResourcePlugin is forcefully activated before do not leads to any issues.

### CODE REVIEW

**Comment by iloveeclipse** [comment]
Interesting change, please give me a chance to review, don't merge too fast. The area of resources init is extremely sensible.

**Comment by laeubi** [comment]
Just let me know if anything is unclear, you can add yourself as a reviewer in the upper right "reviewers" section to indicate that you want review a certain PR.

For testing I used a runconfig without an instance area so I get the Workspace Chooser Dialog, then I added a plugin that simply calls `ResourcesPlugin.getPlugin()`. Now you can set the start level of that testplugin to '1' to make it activate the Resource-Plugin as early as possible.
If you put a debugger on that plugins activator, you see that the plugin gets started, and activates the ResourcesPlugin, then chooser dialog is shown and as soon as you have chosen a location the code to init and publish the workspace is run.

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java
Comment (iloveeclipse) [inline bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java:466]: Could we extract this inner class to something like below?
That would simplify understanding of the code I think.

```
diff --git a/bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java b/bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java
index 6b1c549..49e0a2b 100644
--- a/bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java
+++ b/bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java
@@ -47,2 +47,3 @@
 public final class ResourcesPlugin extends Plugin {
+
 	/**
@@ -468,52 +469,3 @@
 						Location.SERVICE_PROPERTY_URL)),
-				new ServiceTrackerCustomizer<>() {
-
-					private ServiceRegistration<IWorkspace> workspaceRegistration;
-
-					@Override
-					public Workspace addingService(ServiceReference<Location> reference) {
-						if (workspace == null) {
-							Location location = context.getService(reference);
-							if (location != null) {
-								// Remember workspace before opening, to
-								// make it easier to debug cases where open() is failing.
-								workspace = new Workspace();
-								IStatus result;
-								try {
-									result = workspace.open(null);
-									if (!result.isOK())
-										getLog().log(result);
-									workspaceRegistration = context.registerService(IWorkspace.class, workspace, null);
-									return workspace;
-								} catch (CoreException e) {
-									getLog().log(e.getStatus());
-								}
-							}
-						}
-						return null;
-					}
-
-					@Override
-					public void modifiedService(ServiceReference<Location> reference, Workspace service) {
-						// nothing to care about here as we already tracking the service with the
-						// properties of interest
-					}
-
-					@Override
-					public void removedService(ServiceReference<Location> reference, Workspace service) {
-						if (service == workspace) {
-							workspaceRegistration.unregister();
-							try {
-								service.close(null);
-								// Forget workspace only if successfully closed, to
-								// make it easier to debug cases where close() is failing.
-								workspace = null;
-							} catch (CoreException e) {
-								getLog().log(e.getStatus());
-							} finally {
-								context.ungetService(reference);
-							}
-						}
-					}
-				});
+				new WorkspaceInit(context));
 		instanceLocationTracker.open();
@@ -521,2 +473,63 @@
 
+	/**
+	 * Creates and opens singleton workspace instance for this session
+	 */
+	private final class WorkspaceInit implements ServiceTrackerCustomizer<Location, Workspace> {
+		private final BundleContext context;
+		private ServiceRegistration<IWorkspace> workspaceRegistration;
+
+		private WorkspaceInit(BundleContext context) {
+			this.context = context;
+		}
+
+		@Override
+		public Workspace addingService(ServiceReference<Location> reference) {
+			if (workspace != null) {
+				return null;
+			}
+			Location location = context.getService(reference);
+			if (location == null) {
+				return null;
+			}
+			// Remember workspace before opening, to
+			// make it easier to debug cases where open() is failing.
+			workspace = new Workspace();
+			try {
+				IStatus result = workspace.open(null);
+				if (!result.isOK()) {
+					getLog().log(result);
+				}
+				workspaceRegistration = context.registerService(IWorkspace.class, workspace, null);
+				return workspace;
+			} catch (CoreException e) {
+				getLog().log(e.getStatus());
+			}
+			return null;
+		}
+
+		@Override
+		public void modifiedService(ServiceReference<Location> reference, Workspace service) {
+			// nothing to care about here as we already tracking the service with the
+			// properties of interest
+		}
+
+		@Override
+		public void removedService(ServiceReference<Location> reference, Workspace service) {
+			if (service != workspace) {
+				return;
+			}
+			workspaceRegistration.unregister();
+			try {
+				service.close(null);
+				// Forget workspace only if successfully closed, to
+				// make it easier to debug cases where close() is failing.
+				workspace = null;
+			} catch (CoreException e) {
+				getLog().log(e.getStatus());
+			} finally {
+				context.ungetService(reference);
+			}
+		}
+	}
+
 }
```

Comment (laeubi) [reply bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java:468]: Done.

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java
Comment (iloveeclipse) [inline bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java:490]: there is a race condition possible now.
If the client is in getWorkspace() call at line 431, it can get an unitialized workspace, before it is opened.
You can put breakpoint at line 493 to see what I mean.

Comment (laeubi) [reply bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java]: Fixed that.

**Comment by laeubi** [comment]
@iloveeclipse it seems some components in `Workspace#open` rely on calling the static getter... I need to fix that first.

File: tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/resources/ResourceTest.java
Comment (iloveeclipse) [inline tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/resources/ResourceTest.java:86]: Why that? I have no IDE at hand now, but why this trivial getter that is supposed to read a field is replaced with something ... complex and non trivial?

Comment (laeubi) [reply tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/resources/ResourceTest.java:86]: I used that to test if it is simply that the workspace is not ready yet, but its more likely cause by #86

**Comment by laeubi** [comment]
All tests are green here so ready to resume with review, for the remaining issue of storing default charset (will currently issue a warning if creating a new workspace) I have created https://github.com/eclipse-platform/eclipse.platform.resources/pull/87

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java
Comment (iloveeclipse) [inline bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java:417]: I count ~1400 uses of this method in my SDK workspace and ~2600 in my product workspace.
Original code simply did a null check and return. No side effects, not much to worry about.

New code is overwhelmingly complex (for something that is supposed to return a value that is computed once) and adds calls to method that is not recommended to be called on activation (javadoc: `It is strongly recommended that waitForService is not used during the calling of the BundleActivator methods. BundleActivator methods are expected to complete in a short period of time.`). I see that this method **is** called in activators.

By all means, this simple getter shouldn't do any blocking operations or cause other side effects through the framework calls behind. The probability that one of the ~1000 callers will have a side effect (even if this a performance drop) is just too high.

Comment (iloveeclipse) [reply bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java:417]: > So any of the consumers who is scared about the complexity here could simply migrate to consuming the service here

I'm scared because from a simple getter we get a complex code. This method is used a lot everywhere, since day one, and surely no one will migrate anything here to some complex services pattern.

Comment (laeubi) [reply bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java:417]: If you prefer to simply fail I can adjust, beside that a service pattern is nothing "complex", what makes it complex is that most Eclipse plugins don't use anything here but rely on static method calls instead.

Eg I recently refactored [CheckMissingNaturesListener](https://github.com/eclipse-platform/eclipse.platform.resources/blob/master/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/CheckMissingNaturesListener.java) so I don't see it becomes more complex to have
```
@Reference
private IWorkspace workspace;
```
versus
```
ResourcesPlugin.getWorkspace()
```

beside at the latter I have to be scared to "unintentionally" activate the Resource plugin and run into all the trouble with Workspace (not yet) set.

Comment (laeubi) [reply bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java:417]: With https://github.com/eclipse-platform/eclipse.platform.resources/pull/87 I don't see any more calls to getWorkspace while openening, so once this is merged this method could be simplified again.

**Comment by mickaelistria** [comment]
What's missing here before we can merge?

**Comment by laeubi** [comment]
I mainly need some time to finish the remaining "early access" ticket and home that there are not more of that kind :-)

**Comment by laeubi** [comment]
There are now two tests left that seem to hook into the Workspace setup for test-purpose in an early stage but I don't know if that is a valid use case or we should adjust the test cases itself...

**Comment by laeubi** [comment]
> I don't think we can easily dismiss the 2 cases covered by those tests,

I don't think we should dismiss them, but probably the code should be adjusted.

> fail now on early phase as some plugins use it already and it has been working without much issue for a long time.

It fails already today but with the difference that it fails the complete IDE instead of a single plugin

> What you're suggesting with this change results in an API breakage

The APi does not tell anything about startup order and would have failed already today.

> can now cause errors in cases it didn't and the API doesn't tell explicitly what's possible or not

It already throws this exception as of today...

> before we're good to depreacte `ResourcesPlugin.getWorkspace()`.

this won't realistically happen and is not the goal of this change, this should make the ResourcePlugin don't completely fail if other plugins behave wrong, I don't want to change other bad behaving plugins.

**Comment by laeubi** [comment]
> Was this failing with latest release?

If plugin is activate before the workspace is set yes, but with a different exception. These tests now fail, because they access the Workspace before it is initialized (what is not covered with current code) but as this is from within the resource-plugin we should fix this (as was done with the #98 in other parts of the resource plugin).

**Comment by mickaelistria** [comment]
> If plugin is activate before the workspace is set yes, but with a different exception. 

The tests here and the related bug seem to indicate the contrary was intentionally implemented: that it can be possible to access the workspace early and not get a NPE.

**Comment by laeubi** [comment]
@iloveeclipse I have now successfully removed all side-band communication inside `resources` but it seems there is still one place where access to not fully initialized workspace is assumed:

```
at org.eclipse.core.resources.ResourcesPlugin.getWorkspace(ResourcesPlugin.java:423)
at org.eclipse.core.tests.internal.filesystem.bug369177.Bug369177FileSystem.runTestScenario(Bug369177FileSystem.java:42)
at org.eclipse.core.tests.internal.filesystem.bug369177.Bug369177FileSystem.getStore(Bug369177FileSystem.java:59)
at org.eclipse.core.internal.filesystem.InternalFileSystemCore.getStore(InternalFileSystemCore.java:110)
at org.eclipse.core.filesystem.EFS.getStore(EFS.java:479)
at org.eclipse.core.internal.resources.AliasManager.addToLocationsMap(AliasManager.java:368)
at org.eclipse.core.internal.resources.AliasManager.buildLocationsMap(AliasManager.java:406)
at org.eclipse.core.internal.resources.AliasManager.startup(AliasManager.java:611)
at org.eclipse.core.internal.resources.Workspace.startup(Workspace.java:2560)
at org.eclipse.core.internal.resources.Workspace.open(Workspace.java:2251)
```
The corresponding bug is https://bugs.eclipse.org/bugs/show_bug.cgi?id=369177

So even if I strongly agree with you that making partially initialized objects available is bad, I don't see how we can prevent this for this test and there is a chance we break existing codes that use extension points as we have no control over what they call and when they call `ResourcesPlugin.getWorkspace` and need to retain current behavior.

So I think, for any code that want to prevent the possibility to have partially initialized Workspace objects the solution is simply not call `ResourcesPlugin.getWorkspace` but to track the service that is only published if a workspace is fully open and without any critical errors.

**Review by iloveeclipse — CHANGES_REQUESTED** [review]
1) Please rebase.
2) Some small issues inside code, see inline comments
3) I still can break resources plugin init (**without** `-Dosgi.dataAreaRequiresExplicitInit=true`) if Eclipse is started without `-data` by calling `ResourcesPlugin.getEncoding();` from an immediate service.

I think we can add check for initialized workspace to the  `ResourcesPlugin.getEncoding();`

here stack where workspaxce is initiated to default location **while workspace prompt is still shown**
```
java.lang.Exception
	at org.eclipse.core.resources.ResourcesPlugin$WorkspaceInitCustomizer.addingService(ResourcesPlugin.java:492)
	at org.eclipse.core.resources.ResourcesPlugin$WorkspaceInitCustomizer.addingService(ResourcesPlugin.java:1)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:943)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:1)
	at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)
	at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:229)
	at org.osgi.util.tracker.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:903)
	at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:123)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:961)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:234)
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:151)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:937)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:874)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.setProperties(ServiceRegistrationImpl.java:185)
	at org.eclipse.osgi.internal.location.BasicLocation.updateUrl(BasicLocation.java:178)
	at org.eclipse.osgi.internal.location.BasicLocation.set(BasicLocation.java:169)
	at org.eclipse.osgi.internal.location.BasicLocation.set(BasicLocation.java:130)
	at org.eclipse.osgi.internal.location.BasicLocation.setURL(BasicLocation.java:122)
	at org.eclipse.osgi.internal.location.BasicLocation.getURL(BasicLocation.java:97)
	at org.eclipse.core.internal.runtime.DataArea.assertLocationInitialized(DataArea.java:64)
	at org.eclipse.core.internal.runtime.DataArea.getStateLocation(DataArea.java:141)
	at org.eclipse.core.internal.preferences.InstancePreferences.getBaseLocation(InstancePreferences.java:45)
	at org.eclipse.core.internal.preferences.InstancePreferences.initializeChildren(InstancePreferences.java:120)
	at org.eclipse.core.internal.preferences.InstancePreferences.<init>(InstancePreferences.java:60)
	at org.eclipse.core.internal.preferences.InstancePreferences.internalCreate(InstancePreferences.java:131)
	at org.eclipse.core.internal.preferences.EclipsePreferences.create(EclipsePreferences.java:350)
	at org.eclipse.core.internal.preferences.EclipsePreferences.create(EclipsePreferences.java:338)
	at org.eclipse.core.internal.preferences.PreferencesService.createNode(PreferencesService.java:395)
	at org.eclipse.core.internal.preferences.RootPreferences.getChild(RootPreferences.java:63)
	at org.eclipse.core.internal.preferences.RootPreferences.getNode(RootPreferences.java:98)
	at org.eclipse.core.internal.preferences.RootPreferences.node(RootPreferences.java:87)
	at org.eclipse.core.internal.preferences.legacy.PreferenceForwarder.<init>(PreferenceForwarder.java:39)
	at org.eclipse.core.runtime.Plugin.lambda$0(Plugin.java:326)
	at org.eclipse.core.runtime.Plugin.getPluginPreferences(Plugin.java:328)
	at org.eclipse.core.resources.ResourcesPlugin.getEncoding(ResourcesPlugin.java:386)
	at de.loskutov.workspace.StartupComponent$1.run(StartupComponent.java:25)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```

**Comment by laeubi** [comment]
> 3\. I still can break resources plugin init (**without** `-Dosgi.dataAreaRequiresExplicitInit=true`) if Eclipse is started without `-data` by calling `ResourcesPlugin.getEncoding();` from an immediate service.

I have no really good idea here instead of asking people to not use that method and instead call `IWorkspace#getRoot().getDefaultEncoding()` or alike ... but maybe simply throw exception as from `getWorkspace`?

**Comment by iloveeclipse** [comment]
> but maybe simply throw exception as from `getWorkspace`?

That was actually what I've meant, because anyone who accessed encoding **before** workspace location was initialized was most likely using wrong one. 

The main (compatibility) question here is: what if there are applications that didn't care about workspace at all and just always "used" default workspace without ever realizing that some code reads encoding from that? 

I believe yes...

Most prominent "user" of that API is `org.eclipse.jdt.core.JavaCore.getEncoding()` that already has some workarounds to return some default value, see the code and https://bugs.eclipse.org/bugs/show_bug.cgi?id=208807. 
JDT code is often used as formatter / parser / compiler etc in some tools that don't know that JDT actually relies on a workspace to work (also if used as a library).

I think we can do similar: if workspace is not set, return system value **but log an error to the log**. WDYT?

**Comment by laeubi** [comment]
> I think we can do similar: if workspace is not set, return system value **but log an error to the log**. WDYT?

I think this is fine, I must confess I find the method confusing anyways as there seem to be a connection to the workspace but the workspace itself calls this method in some cases, especially with your enhancement to set the workspace encoding on first startup ... so most probably we should even try to get rid of this method in the platform and instead use `org.eclipse.core.resources.IContainer.getDefaultCharset(boolean)` and maybe add an `org.eclipse.core.resources.IContainer.getDefaultCharset()` that returns the system property encoding if the other method would return `null` ...

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java
Comment (laeubi) [reply bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java]: done

File: bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java
Comment (laeubi) [reply bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ResourcesPlugin.java]: done

**Comment by laeubi** [comment]
I have now added some more javadoc and guards to the `getEncoding` method.

**Review by iloveeclipse — APPROVED** [review]
We are one day past the ["official" deadline for M3](https://www.eclipse.org/lists/platform-releng-dev/msg38405.html) : `Monday: Last day of development, (and even then, no "big changes")`.

However, I believe we've made a good progress here, and the "final" change seem to be safe. I think the change is worth to be integrated in M3, because it makes workspace init more robust against possible misuse.

**Comment by iloveeclipse** [comment]
@mickaelistria : do you agree with my last comment?

**Comment by laeubi** [comment]
> I think the change is worth to be integrated in M3, because it makes workspace init more robust against possible misuse.

I'll leave the decision up to you and @mickaelistria here, if we can get it into the next release I'm also happy with that given that this has struggling me for many years now ;-)

**Comment by mickaelistria** [comment]
@akurtakov @laeubi Yes, this is an important change on an area of work that was already started in previous milestones and can be perceived as a fix. So +1 to merge ASAP.

**Pull request merged**
